### PR TITLE
ILC error handling

### PR DESCRIPTION
**Pull request by oleh-momot**

### CODE REVIEW

File: ilc/client/setupErrorHandlers.js [hunk 1]
```diff
@@ -1,7 +1,52 @@
...
+
+const isEssentialOrPrimaryFragment = (fragmentKind) => [KIND_OF_PRIMARY_FRAGMENT, KIND_OF_ESSENTIAL_FRAGMENT].includes(fragmentKind);
+
+export default function ({
+    registryConf,
+    getCurrentPath,
+}) {
+    singleSpa.addErrorHandler((err) => {
+        const errorId = uuidv4();
+        const [appName, slotName] = err.appOrParcelName.split('__at__');
```
[the remark this answers is not in the document]
Comment (StyleT) [reply ilc/client/setupErrorHandlers.js]: Idea for consolidation: create ValueObject for registry config.

File: ilc/client/setupErrorHandlers.js [hunk 2]
```diff
@@ -1,7 +1,52 @@
...
+                })
+                .then((text) => {
+                    document.querySelector('html').innerHTML = text;
+                })
+                .catch((err) => {
+                    alert(err.message);
+                });
+        }
+
+        console.error(JSON.stringify(err));
```
Comment (StyleT) [inline ilc/client/setupErrorHandlers.js]: Here you have HUUGE gap in logic :) You generate errorId here & send it to server, but doesn't log it at client side.

Pls also provide in a comment example of the `console.error(JSON.stringify(err));` output.

File: ilc/client/setupErrorHandlers.js [hunk 3]
```diff
@@ -1,7 +1,52 @@
...
+                        return Promise.reject(new Error('Something went wrong!'));
+                    }
+
+                    return res.text()
+                })
+                .then((text) => {
+                    document.querySelector('html').innerHTML = text;
+                })
+                .catch((err) => {
+                    alert(err.message);
```
Comment (StyleT) [inline ilc/client/setupErrorHandlers.js]: 1. Don't show random `err.message` to user
2. Pls log this error with enough context info

File: ilc/client.js [hunk 4]
```diff
@@ -121,14 +122,23 @@ function getCurrentBasePath() {
...
     prevPath = currentPath;
 
     const path = router.match(window.location.pathname + window.location.search);
     if (currentPath !== null && path.template !== currentPath.template) {
         throw new Error('Base template was changed and I still don\'t know how to handle it :(');
     }
+
     currentPath = path;
+
+    if (!navigator.onLine) {
```
[the remark this answers is not in the document]
Comment (StyleT) [reply ilc/client.js]: What if we went offline during `fetch(`/_ilc/page/500`)`?

File: ilc/templates/error.ejs [hunk 5]
```diff
@@ -0,0 +1,396 @@
+<!DOCTYPE html>
```
Comment (StyleT) [inline ilc/templates/error.ejs]: _For next PR_
I came up with a great idea how we can store this file in a better way:
Add it to the registry as "error template" and fetch at ILC startup to build "pre-heated cache".

File: ilc/server.js [hunk 6]
```diff
@@ -7,21 +10,40 @@ const serveStatic = require('./serveStatic');
...
+    res.status(statusCode);
+    res.render('error', { statusCode: 500, errorId });
+});
 
-app.get('*', (req, res) => {
+app.get('*', async (req, res) => {
     req.headers['x-request-uri'] = req.url; //TODO: to be removed & replaced with routerProps
 
-    return tailor.requestHandler(req, res);
+    await tailor.requestHandler(req, res);
```
Comment (StyleT) [inline ilc/server.js]: why do we need this change?

File: ilc/server.js [hunk 7]
```diff
@@ -7,21 +10,40 @@ const serveStatic = require('./serveStatic');
...
+});
+
+app.use((err, req, res, next) => {
```
Comment (StyleT) [inline ilc/server.js]: You're missing logging here.

File: ilc/server.js [hunk 8]
```diff
@@ -7,21 +10,40 @@ const serveStatic = require('./serveStatic');
...
     console.error('Tailor error:');
     console.error(err);
 });
 
+app.set('view engine', 'ejs');
+app.set('views', './templates');
 
 app.use('/_ilc/', serveStatic(config.get('productionMode')));
 
+app.get('/_ilc/page/500/:errorId', (req, res) => {
```
Comment (StyleT) [inline ilc/server.js]: I think that it's better to pass `errorId` via GET params to make it optional. There may be some cases when we need this page w/o error ID.

File: ilc/server.js [hunk 9]
```diff
@@ -7,21 +10,40 @@ const serveStatic = require('./serveStatic');
...
 app.get('/ping', (req, res) => res.send('pong'));
 
-
 const tailor = tailorFactory(config.get('registry.address'), config.get('cdnUrl'));
 tailor.on('error', (req, err) => {
     console.error('Tailor error:');
     console.error(err);
 });
 
+app.set('view engine', 'ejs');
```
Comment (StyleT) [inline ilc/server.js]: Not sure that we need this. We can just use `ejs` lib in the code below.

File: ilc/tailor/request-handler.js [hunk 10]
```diff
@@ -40,209 +41,210 @@ const INTERNAL_SERVER_ERROR = 'Internal Server Error';
...
-        FORMAT_HTTP_HEADERS,
-        request.headers
-    );
-    const spanOptions = parentSpanContext ? { childOf: parentSpanContext } : {};
-    const span = tracer.startSpan('compose_page', spanOptions);
-    span.addTags({
-        [Tags.HTTP_URL]: request.url,
-        [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_SERVER
-    });
+    return new Promise((resolve, reject) => {
```
Comment (StyleT) [inline ilc/tailor/request-handler.js]: Remove promise from here

Comment (StyleT) [reply ilc/tailor/request-handler.js]: Use events instead

File: ilc/monkey/express-promise.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
+/**
```
Comment (StyleT) [inline ilc/monkey/express-promise.js]: DELETE THIS FILE

File: ilc/server.js (hunk 6 above)
Comment (oleh-momot) [reply ilc/server.js]: It was removed

File: ilc/server.js (hunk 7 above)
Comment (oleh-momot) [reply ilc/server.js]: I added it

File: ilc/tailor/request-handler.js (hunk 10 above)
Comment (oleh-momot) [reply ilc/tailor/request-handler.js]: Promise was removed
It was changed

File: ilc/monkey/express-promise.js (hunk 11 above)
Comment (oleh-momot) [reply ilc/monkey/express-promise.js]: It was removed

File: ilc/client.js [hunk 12]
```diff
@@ -121,14 +122,23 @@ function getCurrentBasePath() {
     return currentPath.basePath;
 }
 
+function getCurrentPath() {
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply ilc/client.js]: Updated

File: ilc/server.js (hunk 9 above)
Comment (oleh-momot) [reply ilc/server.js]: Updated

File: ilc/server.js (hunk 8 above)
Comment (oleh-momot) [reply ilc/server.js]: Updated

File: ilc/client/setupErrorHandlers.js (hunk 2 above)
Comment (oleh-momot) [reply ilc/client/setupErrorHandlers.js]: Updated
Capture

File: ilc/client/setupErrorHandlers.js [hunk 13]
```diff
@@ -1,7 +1,52 @@
+import * as singleSpa from 'single-spa';
+import * as uuidv4 from 'uuid/v4';
+
 const System = window.System;
 const newrelic = window.newrelic;
 
-export default function () {
+const KIND_OF_PRIMARY_FRAGMENT = 'primary';
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply ilc/client/setupErrorHandlers.js]: Updated

File: ilc/client.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (oleh-momot) [reply ilc/client.js]: I moved this check into SPA error handler at the setup error handlers file
I added a cache to initial load 500 error page template from a server

**Comment by StyleT** [comment]
@oleh-momot review #42 ;)

File: ilc/api/apiRegistry.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
+const axios = require('axios');
+const config = require('config');
```
Comment (StyleT) [inline ilc/api/apiRegistry.js]: Inject config & fetchWithCache

File: ilc/api/apiPage500.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
+const fetchWithCacheFactory = require('./fetchWithCache');
+
+const getTemplate = fetchWithCacheFactory(() => {
+    console.log('Calling 500 error page endpoint...');
+
+    return fetch('/_ilc/page/500').then((res) => {
+        if (!res.ok) {
+            return Promise.reject(new Error('Something went wrong!'));
```
Comment (StyleT) [inline ilc/api/apiPage500.js]: More meaningful error pls

File: ilc/server.js [hunk 16]
```diff
@@ -1,29 +1,99 @@
...
+    server(app);
+}
 
+function setupAppHealthCheck() {
+    return new Promise((resolve) => {
+        let intervalId;
 
-app.get('*', (req, res) => {
-    req.headers['x-request-uri'] = req.url; //TODO: to be removed & replaced with routerProps
+        intervalId = setInterval(() => {
```
Comment (StyleT) [inline ilc/server.js]: Remove this interval

File: ilc/api/apiRegistry.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+});
+
+const getTemplateByTemplateNameCreator = ({
+    fetchWithCacheFactory,
+    fetch,
+    registryAddress,
+}) => fetchWithCacheFactory(async (templateName) => {
+    console.log('Calling get template registry endpoint...');
+
+    const res = await fetch(`${registryAddress}/api/v1/template/${encodeURIComponent(templateName)}`, {
```
Comment (StyleT) [inline ilc/api/apiRegistry.js]: use url-join!

File: ilc/api/fetchWithCache.js [hunk 18]
```diff
@@ -0,0 +1,53 @@
+const fetchWithCacheFactory = (fetch, cacheParams) => {
+    const {
+        cacheForSeconds,
+    } = cacheParams;
+
+    let _data = null;
```
Comment (nightnei) [inline ilc/api/fetchWithCache.js]: remove all prefixes "_" in this file

File: ilc/client/setupErrorHandlers.js [hunk 19]
```diff
@@ -1,7 +1,89 @@
...
+});
+
+const selectFragmentKind = (registryConf, path, appName, slotName) => {
+    const appKind = registryConf.apps[appName].kind;
+    const slotKind = path.slots[slotName] && path.slots[slotName].kind;
+
+    return slotKind || appKind;
+}
+
+const isEssentialOrPrimaryFragment = (fragmentKind) => [
```
Comment (nightnei) [inline ilc/client/setupErrorHandlers.js]: we use it only one time. probably it's better to remove function and just call `.includes` inside `if`

File: ilc/client/ApiPage500.js [hunk 20]
```diff
@@ -0,0 +1,67 @@
+export default class ApiPage500 {
+    _cacheForSeconds;
+
+    _endpoint = '/_ilc/page/500';
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply ilc/client/ApiPage500.js]: Updated

File: ilc/api/apiRegistry.js (hunk 14 above)
Comment (oleh-momot) [reply ilc/api/apiRegistry.js]: Updated

File: ilc/server.js (hunk 16 above)
Comment (oleh-momot) [reply ilc/server.js]: Removed

File: ilc/api/apiRegistry.js (hunk 17 above)
Comment (oleh-momot) [reply ilc/api/apiRegistry.js]: Updated

File: ilc/api/fetchWithCache.js (hunk 18 above)
Comment (oleh-momot) [reply ilc/api/fetchWithCache.js]: Removed

**Pull request closed without merging**